### PR TITLE
[Feat] : Support Async chunk cleanup

### PR DESCRIPTION
**Pull request by Sy0307**
## Purpose

To solve #1071, You can see it for more details. 
I add cleanup function in `shm_connector.py ` and cleanup for send/receiver when check_payload_finished is pass.

## Test Plan

New unit-test: `tests/distributed/omni_connectors/test_chunk_transfer_adapter.py`

## Test Result

All tests Passed.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by Sy0307** [comment]
PTAK @R2-Y

File: vllm_omni/distributed/omni_connectors/adapter.py [hunk 1]
```diff
@@ -268,6 +299,9 @@ def get_chunk_for_generation(
     target_stage_id = stage_id - 1
     request_id = request.external_req_id
 
+    if getattr(request, "status", None) == RequestStatus.FINISHED_STOPPED:
+        return
+
     if request_id in connector.finished_requests:
```
Comment (R2-Y) [inline vllm_omni/distributed/omni_connectors/adapter.py]: Because here you already determine whether the request is completed based on its status, putting the request_id into finished_requests seems unnecessary, so this check can be removed.

File: vllm_omni/distributed/omni_connectors/adapter.py [hunk 2]
```diff
@@ -277,9 +311,13 @@ def get_chunk_for_generation(
     if not payload_data:
         return
 
-    if payload_data.get("finished"):
+    if check_payload_finished(payload_data):
         connector.finished_requests.add(request_id)
```
Comment (R2-Y) [inline vllm_omni/distributed/omni_connectors/adapter.py]: same as above

**Comment by R2-Y** [comment]
@amy-why-3459 PTAL

**Comment by hsliuustc0106** [comment]
add test results please

File: vllm_omni/distributed/omni_connectors/adapter.py [hunk 3]
```diff
@@ -368,6 +407,11 @@ def put_chunk(
         if success:
             connector.put_requests[request_id] += 1
             logger.info(f"[Stage-{stage_id}] Sent {connector_put_key}")
+            if check_payload_finished(payload_data):
```
Comment (amy-why-3459) [inline vllm_omni/distributed/omni_connectors/adapter.py]: Perhaps when put_chunk, we can change check_payload_finished to request.is_finished() to reduce the computation.

File: vllm_omni/distributed/omni_connectors/adapter.py [hunk 4]
```diff
@@ -223,9 +239,25 @@ def get_chunk(
...
         if payload_data:
             cached_reqs.additional_information[cached_req_id] = payload_data
-            if payload_data.get("finished"):
+            if check_payload_finished(payload_data):
                 connector.finished_requests.add(req_id)
 
+    # When the *local* stage finishes a request, it is safe to cleanup all
+    # per-request connector state (including finished marker from upstream).
+    finished_req_ids = getattr(scheduler_output, "finished_req_ids", None)
+    if finished_req_ids:
```
Comment (amy-why-3459) [inline vllm_omni/distributed/omni_connectors/adapter.py]: Perhaps we can call clean_up after scheduler._free_request is called, ensuring that all requests can perform the cleanup operation once the request is finished.

**Comment by hsliuustc0106** [comment]
fix DCO

File: vllm_omni/distributed/omni_connectors/adapter.py (hunk 4 above)
Comment (Sy0307) [reply vllm_omni/distributed/omni_connectors/adapter.py]: This way looks better! I will follow in this way.

File: vllm_omni/distributed/omni_connectors/adapter.py (hunk 3 above)
Comment (Sy0307) [reply vllm_omni/distributed/omni_connectors/adapter.py]: If `payload_data["finished"]` always equals `request.is_finished()` ? If so I can change it to request.is_finished as judgement.

File: vllm_omni/distributed/omni_connectors/adapter.py (hunk 1 above)
Comment (Sy0307) [reply vllm_omni/distributed/omni_connectors/adapter.py]: It makes sense. Fixed.

Comment (Sy0307) [reply vllm_omni/distributed/omni_connectors/adapter.py]: It makes sense. Fixed.

File: vllm_omni/distributed/omni_connectors/adapter.py (hunk 3 above)
Comment (amy-why-3459) [reply vllm_omni/distributed/omni_connectors/adapter.py]: They are equal at the put_chunk stage. "finished": torch.tensor(request.is_finished(), dtype=torch.bool)

File: vllm_omni/distributed/omni_connectors/adapter.py [hunk 5]
```diff
@@ -268,7 +300,8 @@ def get_chunk_for_generation(
     target_stage_id = stage_id - 1
     request_id = request.external_req_id
 
-    if request_id in connector.finished_requests:
+    status = getattr(request, "status", None)
+    if status is not None and RequestStatus.is_finished(status):
```
Comment (amy-why-3459) [inline vllm_omni/distributed/omni_connectors/adapter.py]: connector.finished_requests is used to identify that get_chunk is completed. request.states cannot be used. For example, after get_chunk in stage-1 is completed, inference continues for a period of time.

**Comment by Sy0307** [comment]
Solve all CI issues. Please check this PR cc @R2-Y @amy-why-3459 @hsliuustc0106

File: tests/distributed/omni_connectors/test_cleanup.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (hsliuustc0106) [inline tests/distributed/omni_connectors/test_cleanup.py]: @yenuo26 PTAL

File: vllm_omni/core/sched/omni_ar_scheduler.py [hunk 7]
```diff
@@ -66,6 +66,8 @@ def __init__(self, *args, **kwargs):
         self.transfer_triggered_requests: set[str] = set()
```
Comment (hsliuustc0106) [inline vllm_omni/core/sched/omni_ar_scheduler.py:65]: @tzhouam PTAL

Comment (tzhouam) [reply vllm_omni/core/sched/omni_ar_scheduler.py:65]: The scheduler logic looks good to me, though there’s still room for optimization.

Since both the list(set(...)) conversion and the for loop are O(N), we may want to revisit this later if performance becomes a concern.

File: tests/distributed/omni_connectors/test_cleanup.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    assert request_id not in connector.get_requests
+    assert request_id not in connector.finished_requests
+    assert request_id not in connector.request_payload
+    assert request_id not in connector.request_prompt_token_ids
+    assert request_id not in connector.code_prompt_token_ids
+    assert request_id not in connector.request_ids_mapping
+    assert request_id not in connector.request_ids_mapping.values()
+
+    # Idempotent
+    connector.cleanup(request_id)
```
Comment (yenuo26) [inline tests/distributed/omni_connectors/test_cleanup.py]: I think writing it in the following way might better align with the habits of test cases：
    try:
        connector.cleanup(request_id)
    except Exception as e:
        pytest.fail(f"cleanup fail: {e}")

Comment (Sy0307) [reply tests/distributed/omni_connectors/test_cleanup.py]: I have modified. Thanks.

**Comment by Sy0307** [comment]
I have fixed all the issues and I have no idea why buildkite CI are frozen. How to recover CI test and please check this PR is ready or not ;)

cc @hsliuustc0106 @R2-Y

**Comment by yenuo26** [comment]
> I have fixed all the issues and I have no idea why buildkite CI are frozen. How to recover CI test and please check this PR is ready or not ;)

You can @ the corresponding committer to help add the ready label if this PR is ready.

**Comment by R2-Y** [comment]
please leave the e2e test result for qwen3-omni with open & close async_chunk

File: vllm_omni/distributed/omni_connectors/adapter.py [hunk 9]
```diff
@@ -325,12 +345,18 @@ def put_chunk(
             logger.error(f"Failed to use custom_process_input_func for payload extraction: {e}")
 
         if not payload_data:
-            logger.warning("[Stage-%d] No payload data to send for request %s", stage_id, request_id)
-            return
+            # For async_chunk stage1->stage2, we must always deliver a final
+            # finished marker; otherwise Stage-2 will keep polling for a
+            # non-existent SHM segment and hang until CI timeout.
+            if stage_id == 1 and request_finished:
+                payload_data = {"code_predictor_codes": [], "finished": torch.tensor(True, dtype=torch.bool)}
```
Comment (amy-why-3459) [inline vllm_omni/distributed/omni_connectors/adapter.py]: Why do we need to transmit an empty list?

File: vllm_omni/distributed/omni_connectors/adapter.py [hunk 10]
```diff
@@ -346,10 +372,43 @@ def put_chunk(
         if stage_id == 1:
             # TODO: Make parameters configurable and optimize algorithms
             chunk_size = left_context_size = 25
+            codes = payload_data.get("code_predictor_codes", [])
+            if not codes:
+                if not request_finished:
+                    return
+                payload_data.setdefault("finished", torch.tensor(True, dtype=torch.bool))
```
Comment (amy-why-3459) [inline vllm_omni/distributed/omni_connectors/adapter.py]: Why do we need this?

File: vllm_omni/distributed/omni_connectors/adapter.py (hunk 9 above)
Comment (Sy0307) [reply vllm_omni/distributed/omni_connectors/adapter.py]: As long as qwen3 omni Stage-2 does not receive a payload with finished=True, it will keep polling the next key (looping inside get_through_connector()).
However, at certain steps where the upstream is 'finished' but provides no available codes, custom_process_input_func may return None or [] (falsy values), which causes the polling loop to hang.

File: vllm_omni/distributed/omni_connectors/adapter.py (hunk 10 above)
Comment (Sy0307) [reply vllm_omni/distributed/omni_connectors/adapter.py]: Like last reply, to prevent codes size is zero which appeared in qwen3 tts e2e test.

File: vllm_omni/distributed/omni_connectors/adapter.py (hunk 9 above)
Comment (amy-why-3459) [reply vllm_omni/distributed/omni_connectors/adapter.py]: Why does this situation occur?

Comment (amy-why-3459) [reply vllm_omni/distributed/omni_connectors/adapter.py]: When codes is an empty list, we will not send.

File: vllm_omni/distributed/omni_connectors/adapter.py (hunk 5 above)
Comment (amy-why-3459) [reply vllm_omni/distributed/omni_connectors/adapter.py]: This issue has not yet been resolved.

Comment (Sy0307) [reply vllm_omni/distributed/omni_connectors/adapter.py]: Previously, I removed `request.status = RequestStatus.FINISHED_STOPPED` to solve this issue. But I am not sure if  this takes other issues. So I will revert this modification.

File: vllm_omni/distributed/omni_connectors/adapter.py (hunk 9 above)
Comment (Sy0307) [reply vllm_omni/distributed/omni_connectors/adapter.py]: This is exactly where stage2 might get stuck because we don't send an empty list. It keeps waiting because it's expecting a finish signal. I encountered this case while testing Qwen TTS.

Comment (amy-why-3459) [reply vllm_omni/distributed/omni_connectors/adapter.py]: Understood. If the changes are related to TTS, I suggest we conduct a focused review of the subsequent PRs for TTS modifications. In our design, such a situation should not occur. If this does happen, we need to examine whether it is a bug.

File: vllm_omni/core/sched/omni_generation_scheduler.py [hunk 11]
```diff
@@ -354,8 +375,17 @@ def update_from_output(
             routed_experts = None
 
             # Diffusion request: completes in one step; mark finished and free resources
-            if request.status == RequestStatus.FINISHED_STOPPED or (
-                self.omni_connector is None and request.num_computed_tokens >= request.num_prompt_tokens
+            omni_finished_and_consumed = False
+            if self.omni_connector is not None:
+                external_req_id = getattr(request, "external_req_id", None)
+                if external_req_id is not None:
+                    omni_finished_and_consumed = external_req_id in getattr(
```
Comment (amy-why-3459) [inline vllm_omni/core/sched/omni_generation_scheduler.py]: Why do we need this?

**Comment by R2-Y** [comment]
I tested this pr with async chunk on qwen3-omni, test passed.

**Comment by hsliuustc0106** [comment]
fix CI

**Comment by Sy0307** [comment]
> fix CI

Focus on Qwen3 tts work now. I will fix it later.

**Comment by Gaohan123** [comment]
Any updates? Thanks

**Comment by Sy0307** [comment]
> Any updates? Thanks

Sorry, I am working on tiny issues on #1161 recent days and I will fix CI issue when this part of work is done, maybe tomorrow.

**Comment by lishunyang12** [comment]
@Sy0307 Hey, the shm_connector cleanup for async chunks fixes #1071 — without this, shared memory resources could leak between send/receive cycles. You mentioned the test plan is still TODO. Have you been able to add tests for this?

**Comment by Sy0307** [comment]
> @Sy0307 Hey, the shm_connector cleanup for async chunks fixes #1071 — without this, shared memory resources could leak between send/receive cycles. You mentioned the test plan is still TODO. Have you been able to add tests for this?

As #951 changed the structure of async chunk (ref "Refactor SharedMemoryConnector"), I need to adjust the content of this PR and after refactor I will offer more details of test plan. Test plan will contain new unit-tests and may add checkpoint for e2e test.

**Comment by Sy0307** [comment]
PTAK code after refactor cc @hsliuustc0106 @R2-Y @lishunyang12

**Comment by amy-why-3459** [comment]
@Gaohan123 @R2-Y @Shirley125 @hsliuustc0106 PTAL

**Comment by R2-Y** [comment]
LGTM, thanks for your contribution

File: vllm_omni/model_executor/stage_input_processors/qwen3_tts.py [hunk 12]
```diff
@@ -40,27 +47,25 @@ def talker2code2wav_async_chunk(
...
+    length = len(transfer_manager.code_prompt_token_ids[request_id])
 
-    frame = _extract_last_frame(pooling_output)
-    if frame is not None:
-        codec_codes = frame.cpu().tolist()
-        transfer_manager.code_prompt_token_ids[request_id].append(codec_codes)
+    # Avoid emitting empty chunks during normal streaming. If the request is
+    # finished and nothing was produced, emit an EOF marker.
+    if length <= 0:
+        if finished:
```
Comment (amy-why-3459) [inline vllm_omni/model_executor/stage_input_processors/qwen3_tts.py]: Under what circumstances would a request complete without generating a code? Does this mean we generated empty audio?

**Comment by Sy0307** [comment]
**Before #1378**: save_async synchronously called custom_process_next_stage_input_func on the main thread to read code_prompt_token_ids, meaning fully processed payload_data was enqueued. Executing cleanup after save_async was safe.

**After #1378**: save_async only enqueues raw data (pooling_output, request), and payload processing is deferred to _send_single_request on a background thread. This creates a risk where the main thread's cleanup clears the state before the background thread can read it.

To address this, I added a new parameter defer_cleanup to accommodate this asynchronous logic. Please refer to commit 438331d.

cc @amy-why-3459

**Comment by amy-why-3459** [comment]
Thank you so much for your contribution, it's great work. However, since `cleanup` is called after the `finished` request, it's unnecessary to call `cleanup` in `chunk_transfer_adapter.py`. I tested your branch and found that `cleanup` was called three times per request. After removing the call in `chunk_transfer_adapter.py`, it returned to normal, with only one call per request.

**Comment by amy-why-3459** [comment]
nice work, LGTM.
